File: sim/wheat_field.go
package sim

import (
	"fmt"
	"math"
	"sync"

	"github.com/nelsong6/ambience/rngutil"
)

type WheatFieldConfig = ProceduralConfig
type WheatFieldState = ProceduralState
type WheatFieldSnapshot = ProceduralSnapshot
type WheatFieldPersistedState = ProceduralPersistedState

// WheatField is a dedicated sim type for the wheat-field effect, replacing
// the kind="wheat-field" branch of the old Procedural type.
type WheatField struct {
	mu sync.Mutex

	W, H int
	Grid [][]Pixel

	rng    *rngutil.RNG
	cfg    WheatFieldConfig
	tick   int
	timers map[string]int
	values map[string]float64
	log    []LogEntry
}

var wheatFieldDefaultsLocal = WheatFieldConfig{
	"intro_dur":     60,
	"intro_breeze":  0.16,
	"ending_dur":    70,
	"ending_linger": 20,
	"ending_sway":   0.08,
	"density":       0.48,
	"speed":         0.12,
	"drift":         0.16,
	"sway":          0.68,
	"wave_freq":     0.18,
	"field_top":     0.62,
	"stalk_h":       18,
	"layers":        3,
	"hue":           46,
	"hue_sp":        18,
	"sat":           0.64,
	"lmin":          0.30,
	"lmax":          0.76,
	"gust_p":        0.0,
	"calm_p":        0.0,
	"gust_dur":      50,
	"gust_mult":     1.85,
	"calm_dur":      72,
	"calm_mult":     0.40,
}

func WheatFieldSchema() EffectSchema {
	return EffectSchema{
		Name: "wheat-field",
		Knobs: []Knob{
			{Key: "intro_dur", Label: "intro dur", Slot: SlotSpawn, Group: "introduction", Type: KnobInt, Min: 10, Max: 240, Step: 5, Default: 60, Trigger: "intro",
				Description: "Ticks spent spreading motion through the field from near-stillness into full waves."},
			{Key: "intro_breeze", Label: "intro breeze", Slot: SlotSpawn, Group: "introduction", Type: KnobFloat, Min: 0.08, Max: 0.5, Step: 0.02, Default: 0.16,
				Description: "Starting fraction of the full sway before the waves finish arriving."},
			{Key: "ending_dur", Label: "ending dur", Slot: SlotEnd, Group: "ending", Type: KnobInt, Min: 10, Max: 240, Step: 5, Default: 70, Trigger: "ending",
				Description: "Ticks spent damping the field back toward calm."},
			{Key: "ending_linger", Label: "ending linger", Slot: SlotEnd, Group: "ending", Type: KnobInt, Min: 0, Max: 140, Step: 5, Default: 20,
				Description: "Extra quiet ticks for the last residual motion to settle out."},
			{Key: "ending_sway", Label: "ending sway", Slot: SlotEnd, Group: "ending", Type: KnobFloat, Min: 0.04, Max: 0.28, Step: 0.02, Default: 0.08,
				Description: "Residual sway fraction that remains near the end of the outro."},
			{Key: "density", Label: "density", Slot: SlotLever, Group: "field", Type: KnobFloat, Min: 0.24, Max: 0.92, Step: 0.02, Default: 0.48,
				Description: "How densely the field is packed with visible stalk highlights."},
			{Key: "speed", Label: "wave speed", Slot: SlotLever, Group: "field", Type: KnobFloat, Min: 0.02, Max: 0.4, Step: 0.01, Default: 0.12,
				Description: "How quickly the broad waves travel through the field."},
			{Key: "drift", Label: "drift", Slot: SlotLever, Group: "field", Type: KnobFloat, Min: -0.5, Max: 0.5, Step: 0.01, Default: 0.16,
				Description: "Preferred direction of the wave travel. Positive values push right."},
			{Key: "sway", Label: "sway", Slot: SlotLever, Group: "field", Type: KnobFloat, Min: 0.25, Max: 1.35, Step: 0.02, Default: 0.68,
				Description: "How far the stalk tips lean and recover."},
			{Key: "wave_freq", Label: "wave freq", Slot: SlotLever, Group: "field", Type: KnobFloat, Min: 0.06, Max: 0.35, Step: 0.01, Default: 0.18,
				Description: "Horizontal frequency of the passing field waves."},
			{Key: "field_top", Label: "field top", Slot: SlotLever, Group: "field", Type: KnobFloat, Min: 0.54, Max: 0.74, Step: 0.01, Default: 0.62,
				Description: "Where the top of the wheat band sits in the frame."},
			{Key: "stalk_h", Label: "stalk height", Slot: SlotLever, Group: "field", Type: KnobFloat, Min: 8, Max: 28, Step: 1, Default: 18,
				Description: "Apparent height of the stalks rising above the field base."},
			{Key: "layers", Label: "layers", Slot: SlotLever, Group: "field", Type: KnobInt, Min: 1, Max: 4, Step: 1, Default: 3,
				Description: "Number of depth layers used to build the field."},
			{Key: "hue", Label: "hue", Slot: SlotLever, Group: "color", Type: KnobFloat, Min: 38, Max: 66, Step: 1, Default: 46,
				Description: "Base wheat hue. Lower values warm toward amber; higher values lean green-gold."},
			{Key: "hue_sp", Label: "hue spread", Slot: SlotLever, Group: "color", Type: KnobFloat, Min: 0, Max: 32, Step: 1, Default: 18,
				Description: "Variation across stalk highlights and shadow bands."},
			{Key: "sat", Label: "saturation", Slot: SlotLever, Group: "color", Type: KnobFloat, Min: 0.32, Max: 0.95, Step: 0.01, Default: 0.64,
				Description: "Overall color saturation of the field."},
			{Key: "lmin", Label: "light min", Slot: SlotLever, Group: "color", Type: KnobFloat, Min: 0.18, Max: 0.55, Step: 0.01, Default: 0.30,
				Description: "Minimum lightness used for deeper shadowed wheat."},
			{Key: "lmax", Label: "light max", Slot: SlotLever, Group: "color", Type: KnobFloat, Min: 0.55, Max: 0.92, Step: 0.01, Default: 0.76,
				Description: "Maximum lightness used for sunstruck stalk tips."},
			{Key: "gust_p", Label: "gust", Slot: SlotEvent, Type: KnobFloat, Min: 0, Max: 0.02, Step: 0.0005, Default: 0, Trigger: "gust",
				Description: "Per-tick chance of a stronger wind wave crossing the field."},
			{Key: "calm_p", Label: "calm", Slot: SlotEvent, Type: KnobFloat, Min: 0, Max: 0.02, Step: 0.0005, Default: 0, Trigger: "calm",
				Description: "Per-tick chance of the field settling into a quieter, lighter sway."},
			{Key: "gust_dur", Label: "gust dur", Slot: SlotEventMod, Group: "gust", Type: KnobInt, Min: 10, Max: 220, Step: 5, Default: 50,
				Description: "Typical duration of a stronger wind pulse."},
			{Key: "gust_mult", Label: "gust x", Slot: SlotEventMod, Group: "gust", Type: KnobFloat, Min: 1.05, Max: 4, Step: 0.05, Default: 1.85,
				Description: "Sway multiplier applied during a gust."},
			{Key: "calm_dur", Label: "calm dur", Slot: SlotEventMod, Group: "calm", Type: KnobInt, Min: 10, Max: 220, Step: 5, Default: 72,
				Description: "Duration of the quieter low-amplitude window."},
			{Key: "calm_mult", Label: "calm x", Slot: SlotEventMod, Group: "calm", Type: KnobFloat, Min: 0.05, Max: 1, Step: 0.05, Default: 0.4,
				Description: "Sway multiplier applied while calm is active."},
		},
	}
}

func defaultWheatFieldConfig() WheatFieldConfig { return cloneConfig(wheatFieldDefaultsLocal) }

func mergeWheatFieldDefaults(cfg WheatFieldConfig) WheatFieldConfig {
	out := defaultWheatFieldConfig()
	for k, v := range cfg {
		out[k] = v
	}
	if out["intro_dur"] <= 0 {
		out["intro_dur"] = wheatFieldDefaultsLocal["intro_dur"]
	}
	out["intro_breeze"] = clamp01(out["intro_breeze"])
	if out["ending_dur"] <= 0 {
		out["ending_dur"] = wheatFieldDefaultsLocal["ending_dur"]
	}
	if out["ending_linger"] < 0 {
		out["ending_linger"] = 0
	}
	out["ending_sway"] = clamp01(out["ending_sway"])
	if out["density"] <= 0 {
		out["density"] = wheatFieldDefaultsLocal["density"]
	}
	if out["speed"] <= 0 {
		out["speed"] = wheatFieldDefaultsLocal["speed"]
	}
	if out["sway"] <= 0 {
		out["sway"] = wheatFieldDefaultsLocal["sway"]
	}
	if out["wave_freq"] <= 0 {
		out["wave_freq"] = wheatFieldDefaultsLocal["wave_freq"]
	}
	if out["field_top"] <= 0 {
		out["field_top"] = wheatFieldDefaultsLocal["field_top"]
	}
	if out["stalk_h"] <= 0 {
		out["stalk_h"] = wheatFieldDefaultsLocal["stalk_h"]
	}
	if out["layers"] < 1 {
		out["layers"] = wheatFieldDefaultsLocal["layers"]
	}
	if out["hue"] == 0 {
		out["hue"] = wheatFieldDefaultsLocal["hue"]
	}
	if out["hue_sp"] < 0 {
		out["hue_sp"] = 0
	}
	if out["sat"] <= 0 {
		out["sat"] = wheatFieldDefaultsLocal["sat"]
	}
	if out["lmin"] <= 0 {
		out["lmin"] = wheatFieldDefaultsLocal["lmin"]
	}
	if out["lmax"] <= 0 {
		out["lmax"] = wheatFieldDefaultsLocal["lmax"]
	}
	if out["lmax"] < out["lmin"] {
		out["lmin"], out["lmax"] = out["lmax"], out["lmin"]
	}
	if out["gust_dur"] <= 0 {
		out["gust_dur"] = wheatFieldDefaultsLocal["gust_dur"]
	}
	if out["gust_mult"] <= 0 {
		out["gust_mult"] = wheatFieldDefaultsLocal["gust_mult"]
	}
	if out["calm_dur"] <= 0 {
		out["calm_dur"] = wheatFieldDefaultsLocal["calm_dur"]
	}
	if out["calm_mult"] <= 0 {
		out["calm_mult"] = wheatFieldDefaultsLocal["calm_mult"]
	}
	return out
}

func NewWheatField(w, h int, seed int64, cfg WheatFieldConfig) *WheatField {
	grid := make([][]Pixel, h)
	for i := range grid {
		grid[i] = make([]Pixel, w)
	}
	return &WheatField{
		W:      w,
		H:      h,
		Grid:   grid,
		rng:    rngutil.New(seed),
		cfg:    mergeWheatFieldDefaults(cfg),
		timers: make(map[string]int),
		values: make(map[string]float64),
	}
}

func (s *WheatField) Resize(w, h int) {
	if w <= 0 || h <= 0 {
		return
	}
	s.mu.Lock()
	defer s.mu.Unlock()
	if w == s.W && h == s.H {
		return
	}
	s.W = w
	s.H = h
	s.Grid = make([][]Pixel, h)
	for i := range s.Grid {
		s.Grid[i] = make([]Pixel, w)
	}
}

func (s *WheatField) SetConfig(cfg WheatFieldConfig) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.cfg = mergeWheatFieldDefaults(cfg)
}

func (s *WheatField) EffectiveConfig() WheatFieldConfig {
	s.mu.Lock()
	defer s.mu.Unlock()
	return cloneConfig(s.cfg)
}

func (s *WheatField) Snapshot() WheatFieldSnapshot {
	s.mu.Lock()
	defer s.mu.Unlock()
	return WheatFieldSnapshot{ProceduralState: s.snapshotStateLocked()}
}

func (s *WheatField) RestoreSnapshot(snap WheatFieldSnapshot) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.restoreStateLocked(snap.ProceduralState)
}

func (s *WheatField) SnapshotPersistedState() WheatFieldPersistedState {
	s.mu.Lock()
	defer s.mu.Unlock()
	return WheatFieldPersistedState{
		ProceduralState: s.snapshotStateLocked(),
		RNGState:        s.rng.State(),
	}
}

func (s *WheatField) RestorePersistedState(ps WheatFieldPersistedState) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.restoreStateLocked(ps.ProceduralState)
	if ps.RNGState != 0 {
		s.rng.SetState(ps.RNGState)
	}
}

func (s *WheatField) snapshotStateLocked() ProceduralState {
	return ProceduralState{
		Tick:   s.tick,
		Timers: cloneTimerMap(s.timers),
		Values: cloneValueMap(s.values),
	}
}

func (s *WheatField) restoreStateLocked(state ProceduralState) {
	s.tick = state.Tick
	s.timers = cloneTimerMap(state.Timers)
	if s.timers == nil {
		s.timers = make(map[string]int)
	}
	s.values = cloneValueMap(state.Values)
	if s.values == nil {
		s.values = make(map[string]float64)
	}
}

func (s *WheatField) CurrentTick() int {
	s.mu.Lock()
	defer s.mu.Unlock()
	return s.tick
}

func (s *WheatField) PerturbRNG(delta int64) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.rng.Mix(delta)
}

func (s *WheatField) DrainLog() []LogEntry {
	s.mu.Lock()
	defer s.mu.Unlock()
	if len(s.log) == 0 {
		return nil
	}
	out := s.log
	s.log = nil
	return out
}

func (s *WheatField) appendLog(kind, desc string) {
	s.log = append(s.log, LogEntry{Tick: s.tick, Type: kind, Desc: desc})
	if len(s.log) > 200 {
		s.log = s.log[len(s.log)-200:]
	}
}

func (s *WheatField) intCfg(key string) int {
	return int(math.Round(s.cfg[key]))
}

func (s *WheatField) TriggerEvent(name string) bool {
	s.mu.Lock()
	defer s.mu.Unlock()
	switch name {
	case "gust":
		s.startGustLocked("triggered")
	case "calm":
		s.startCalmLocked("triggered")
	case "intro":
		s.startIntroLocked()
		s.appendLog("intro", fmt.Sprintf("started (dur=%d, breeze=%.2f)", s.timers["intro"], s.cfg["intro_breeze"]))
	case "ending":
		s.startEndingLocked()
		s.appendLog("ending", fmt.Sprintf("started (fade=%d, linger=%d)", s.intCfg("ending_dur"), s.intCfg("ending_linger")))
	default:
		return false
	}
	return true
}

func (s *WheatField) Step() {
	s.mu.Lock()
	defer s.mu.Unlock()

	s.tick++
	for key, value := range s.timers {
		if value > 0 {
			s.timers[key] = value - 1
		}
	}
	s.stepLocked()
}

func (s *WheatField) startGustLocked(verb string) {
	s.timers["gust"] = jitterInt(s.rng, s.intCfg("gust_dur"), 0.3)
	sign := 1.0
	if s.rng.Float64() < 0.35 {
		sign = -1
	}
	s.values["gust_push"] = sign * s.cfg["gust_mult"] * (0.55 + s.rng.Float64()*0.55)
	s.appendLog("gust", fmt.Sprintf("%s (dur=%d, push=%+.2f)", verb, s.timers["gust"], s.values["gust_push"]))
}

func (s *WheatField) startCalmLocked(verb string) {
	s.timers["calm"] = jitterInt(s.rng, s.intCfg("calm_dur"), 0.3)
	s.appendLog("calm", fmt.Sprintf("%s (dur=%d, x%.2f)", verb, s.timers["calm"], s.cfg["calm_mult"]))
}

func (s *WheatField) startIntroLocked() {
	s.timers["gust"] = 0
	s.timers["calm"] = 0
	s.timers["ending"] = 0
	s.values["gust_push"] = 0
	s.timers["intro"] = s.intCfg("intro_dur")
	s.values["intro_total"] = float64(s.timers["intro"])
}

func (s *WheatField) startEndingLocked() {
	s.timers["intro"] = 0
	s.timers["gust"] = 0
	s.timers["calm"] = 0
	s.values["gust_push"] = 0
	endingTotal := s.intCfg("ending_dur") + max(0, s.intCfg("ending_linger"))
	if endingTotal < 1 {
		endingTotal = max(1, s.intCfg("ending_dur"))
	}
	s.timers["ending"] = endingTotal
	s.values["ending_total"] = float64(endingTotal)
}

func (s *WheatField) stepLocked() {
	if s.timers["gust"] <= 0 {
		s.values["gust_push"] = 0
	}
	if s.timers["intro"] <= 0 {
		delete(s.values, "intro_total")
	}
	if s.timers["ending"] <= 0 {
		delete(s.values, "ending_total")
	}
	if s.timers["intro"] > 0 || s.timers["ending"] > 0 {
		return
	}
	if s.timers["gust"] <= 0 && s.cfg["gust_p"] > 0 && s.rng.Float64() < s.cfg["gust_p"] {
		s.startGustLocked("started")
	}
	if s.timers["calm"] <= 0 && s.cfg["calm_p"] > 0 && s.rng.Float64() < s.cfg["calm_p"] {
		s.startCalmLocked("started")
	}
}

File: cmd/ambience/web/effects/wheat_field.js
'use strict';
(function (api) {
	const { makeRNG, jitterInt, clamp01, hslToRGB, positiveMod } = api._helpers;

	const DEFAULTS = {
		intro_dur: 60,
		intro_breeze: 0.16,
		ending_dur: 70,
		ending_linger: 20,
		ending_sway: 0.08,
		density: 0.48,
		speed: 0.12,
		drift: 0.16,
		sway: 0.68,
		wave_freq: 0.18,
		field_top: 0.62,
		stalk_h: 18,
		layers: 3,
		hue: 46,
		hue_sp: 18,
		sat: 0.64,
		lmin: 0.3,
		lmax: 0.76,
		gust_p: 0,
		calm_p: 0,
		gust_dur: 50,
		gust_mult: 1.85,
		calm_dur: 72,
		calm_mult: 0.4,
	};

	function applyDefaults(cfg) {
		const c = Object.assign({}, DEFAULTS, cfg || {});
		if (c.intro_dur <= 0) c.intro_dur = DEFAULTS.intro_dur;
		c.intro_breeze = clamp01(c.intro_breeze);
		if (c.ending_dur <= 0) c.ending_dur = DEFAULTS.ending_dur;
		if (c.ending_linger < 0) c.ending_linger = 0;
		c.ending_sway = clamp01(c.ending_sway);
		if (c.density <= 0) c.density = DEFAULTS.density;
		if (c.speed <= 0) c.speed = DEFAULTS.speed;
		if (c.sway <= 0) c.sway = DEFAULTS.sway;
		if (c.wave_freq <= 0) c.wave_freq = DEFAULTS.wave_freq;
		if (c.field_top <= 0) c.field_top = DEFAULTS.field_top;
		if (c.stalk_h <= 0) c.stalk_h = DEFAULTS.stalk_h;
		if (c.layers < 1) c.layers = DEFAULTS.layers;
		if (c.hue === 0) c.hue = DEFAULTS.hue;
		if (c.hue_sp < 0) c.hue_sp = 0;
		if (c.sat <= 0) c.sat = DEFAULTS.sat;
		if (c.lmin <= 0) c.lmin = DEFAULTS.lmin;
		if (c.lmax <= 0) c.lmax = DEFAULTS.lmax;
		if (c.lmax < c.lmin) [c.lmin, c.lmax] = [c.lmax, c.lmin];
		if (c.gust_dur <= 0) c.gust_dur = DEFAULTS.gust_dur;
		if (c.gust_mult <= 0) c.gust_mult = DEFAULTS.gust_mult;
		if (c.calm_dur <= 0) c.calm_dur = DEFAULTS.calm_dur;
		if (c.calm_mult <= 0) c.calm_mult = DEFAULTS.calm_mult;
		return c;
	}

	class WheatField {
		constructor(w, h, cfg, seed) {
			this.w = w;
			this.h = h;
			this.grid = new Uint8ClampedArray(w * h * 3);
			this.seed = Number(seed || Date.now());
			this.tick = 0;
			this.timers = {};
			this.values = {};
			this.cfg = applyDefaults(cfg);
		}

		setConfig(cfg) {
			this.cfg = applyDefaults(Object.assign({}, this.cfg, cfg));
		}

		restoreSnapshot(snap) {
			const state = snap.state || snap;
			this.setConfig(snap.config || {});
			this.tick = state.tick || snap.tick || 0;
			this.timers = Object.assign({}, state.timers || {});
			this.values = Object.assign({}, state.values || {});
			if (typeof snap.seed === 'number') this.seed = snap.seed;
			if (snap.gridW > 0 && snap.gridH > 0) {
				this.w = snap.gridW;
				this.h = snap.gridH;
			}
		}

		_eventRng(salt) {
			return makeRNG(((this.seed >>> 0) ^ (((this.tick + salt) * 2654435761) >>> 0)) >>> 0);
		}

		_hash(index) {
			const x = Math.sin((this.seed * 0.000001 + index * 12.9898) * 43758.5453);
			return x - Math.floor(x);
		}

		_phaseProgress(total, left) {
			if (left <= 1 || total <= 1) return 1;
			const elapsed = total - left;
			if (elapsed <= 0) return 0;
			return clamp01(elapsed / Math.max(1, total - 1));
		}

		_fillCell(ctx, sx, sy, ceilSx, ceilSy, x, y, w, h, color, alpha) {
			ctx.fillStyle = color;
			ctx.globalAlpha = alpha == null ? 1 : alpha;
			ctx.fillRect(Math.floor(x * sx), Math.floor(y * sy), Math.max(1, Math.ceil(w * sx || ceilSx)), Math.max(1, Math.ceil(h * sy || ceilSy)));
			ctx.globalAlpha = 1;
		}

		triggerEvent(name) {
			const rng = this._eventRng(name.length + 47);
			switch (name) {
				case 'gust':
					this.timers.gust = jitterInt(rng, this.cfg.gust_dur, 0.3);
					this.values.gust_push = (rng() < 0.35 ? -1 : 1) * this.cfg.gust_mult * (0.55 + rng() * 0.55);
					return true;
				case 'calm':
					this.timers.calm = jitterInt(rng, this.cfg.calm_dur, 0.3);
					return true;
				case 'intro':
					this.timers.gust = 0;
					this.timers.calm = 0;
					this.timers.ending = 0;
					this.values.gust_push = 0;
					this.timers.intro = Math.max(1, Math.round(this.cfg.intro_dur));
					this.values.intro_total = this.timers.intro;
					return true;
				case 'ending':
					this.timers.intro = 0;
					this.timers.gust = 0;
					this.timers.calm = 0;
					this.values.gust_push = 0;
					this.timers.ending = Math.max(1, Math.round(this.cfg.ending_dur + Math.max(0, this.cfg.ending_linger)));
					this.values.ending_total = this.timers.ending;
					return true;
			}
			return false;
		}

		step() {
			this.tick++;
			for (const key of Object.keys(this.timers)) {
				if (this.timers[key] > 0) this.timers[key]--;
			}
			if (!this.timers.gust || this.timers.gust <= 0) this.values.gust_push = 0;
			api._helpers.paintProceduralGrid(this);
		}

		_motionLevel() {
			let level = this.cfg.sway;
			if (this.timers.gust > 0) level *= 1 + Math.abs(this.values.gust_push || this.cfg.gust_mult) * 0.35;
			if (this.timers.calm > 0) level *= this.cfg.calm_mult;
			if (this.timers.intro > 0) {
				const total = this.values.intro_total || this.cfg.intro_dur;
				const progress = this._phaseProgress(total, this.timers.intro);
				level *= this.cfg.intro_breeze + (1 - this.cfg.intro_breeze) * progress;
			}
			if (this.timers.ending > 0) {
				const total = this.values.ending_total || (this.cfg.ending_dur + this.cfg.ending_linger);
				const progress = this._phaseProgress(total, this.timers.ending);
				level *= 1 - (1 - this.cfg.ending_sway) * progress;
			}
			return Math.max(0.05, level);
		}

		render(ctx, canvasW, canvasH, opts) {
			api._helpers.renderPixelGridEffect(this, ctx, canvasW, canvasH, opts);
		}
	}

	api.presets['wheat-field'] = [
		{
			key: 'still-evening',
			label: 'still evening',
			config: {
				density: 0.4,
				speed: 0.07,
				drift: 0.05,
				sway: 0.34,
				wave_freq: 0.14,
				field_top: 0.64,
				stalk_h: 17,
				layers: 2,
				hue: 42,
				hue_sp: 12,
				sat: 0.56,
				lmin: 0.28,
				lmax: 0.7,
				calm_p: 0.001,
			},
		},
		{
			key: 'gentle-breeze',
			label: 'gentle breeze',
			config: {
				density: 0.48,
				speed: 0.12,
				drift: 0.14,
				sway: 0.68,
				wave_freq: 0.18,
				field_top: 0.62,
				stalk_h: 18,
				layers: 3,
				hue: 46,
				hue_sp: 18,
				sat: 0.64,
				lmin: 0.3,
				lmax: 0.76,
				gust_p: 0.0008,
			},
		},
		{
			key: 'rolling-field',
			label: 'rolling field',
			config: {
				density: 0.56,
				speed: 0.16,
				drift: 0.2,
				sway: 0.88,
				wave_freq: 0.16,
				field_top: 0.6,
				stalk_h: 20,
				layers: 3,
				hue: 48,
				hue_sp: 20,
				sat: 0.68,
				lmin: 0.3,
				lmax: 0.8,
				gust_p: 0.0012,
				gust_mult: 2.15,
			},
		},
		{
			key: 'windy-harvest',
			label: 'windy harvest',
			config: {
				density: 0.62,
				speed: 0.2,
				drift: 0.28,
				sway: 1.02,
				wave_freq: 0.21,
				field_top: 0.59,
				stalk_h: 22,
				layers: 4,
				hue: 44,
				hue_sp: 24,
				sat: 0.72,
				lmin: 0.32,
				lmax: 0.84,
				gust_p: 0.0016,
				gust_mult: 2.45,
				gust_dur: 66,
			},
		},
	];
	api.effects['wheat-field'] = WheatField;
})(window.AmbienceSim);
